File: c/2023/src/day1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "array.h"
#include "commons.h"
#include "string.h"

int part1(StringArray *lines) {
  int result = 0;

  string_array_foreach (char *line, lines) {
    size_t line_len = strlen(line);

    Array *digits = array_new(int);
    for (size_t i = 0; i < line_len; i++) {
      int value = line[i] - '0';
      if (value > 0 && value < 10) {
        array_push_rval(digits, value);
      }
    }
    debugf(int_array_print, digits);
    result += int_array_first(digits) * 10 + int_array_last(digits);

    array_free(digits);
  }

  return result;
}

int part2(StringArray *lines) {
  int result = 0;

  string_array_foreach (char *line, lines) {
    size_t line_len = strlen(line);
    debug("%s: %zu\n", line, line_len);

    Array *digits = array_new(int);
    for (size_t i = 0; i < line_len; i++) {
      char c = line[i];
      if (c >= '1' && c <= '9') {
        array_push_rval(digits, c - '0');
      } else {
        if (strncmp(line + i, "one", 3) == 0) array_push_rval(digits, 1);
        else if (strncmp(line + i, "two", 3) == 0) array_push_rval(digits, 2);
        else if (strncmp(line + i, "three", 5) == 0) array_push_rval(digits, 3);
        else if (strncmp(line + i, "four", 4) == 0) array_push_rval(digits, 4);
        else if (strncmp(line + i, "five", 4) == 0) array_push_rval(digits, 5);
        else if (strncmp(line + i, "six", 3) == 0) array_push_rval(digits, 6);
        else if (strncmp(line + i, "seven", 5) == 0) array_push_rval(digits, 7);
        else if (strncmp(line + i, "eight", 5) == 0) array_push_rval(digits, 8);
        else if (strncmp(line + i, "nine", 4) == 0) array_push_rval(digits, 9);
      }
    }
    debugf(int_array_print, digits);
    result += int_array_first(digits) * 10 + int_array_last(digits);
    array_free(digits);
  }

  return result;
}

int main(void) {
  test_case(day1, part1, sample1, 142);
  test_case(day1, part1, data, 55488);
  test_case(day1, part2, sample2, 281);
  test_case(day1, part2, data, 55614);

  return 0;
}

File: rust/2023/src/bin/day1.rs
use std::collections::HashMap;

static SAMPLE1: &str = include_str!("../../../../inputs/2023/day1/sample1");
static SAMPLE2: &str = include_str!("../../../../inputs/2023/day1/sample2");
static DATA: &str = include_str!("../../../../inputs/2023/day1/data");

fn part1(input: &str) -> u64 {
    let mut sum: u64 = 0;
    for line in input.split('\n').filter(|&line| !line.is_empty()) {
        let mut digits = Vec::new();
        for c in line.chars() {
            if c.is_ascii_digit() {
                digits.push(c.to_digit(10).unwrap() as u64);
            }
        }
        sum += digits.first().unwrap() * 10 + digits.last().unwrap();
    }

    sum
}

fn part2(input: &str) -> u64 {
    let cases = HashMap::from([
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9),
    ]);

    let mut sum: u64 = 0;
    for line in input.split('\n').filter(|&line| !line.is_empty()) {
        let mut digits = Vec::new();
        for (i, c) in line.chars().enumerate() {
            if c.is_ascii_digit() {
                digits.push(c.to_digit(10).unwrap() as u64);
            } else {
                for (case, digit) in cases.iter() {
                    if i + case.len() <= line.len() {
                        let substring = &line[i..i + case.len()];
                        if substring == *case {
                            digits.push(*digit);
                        }
                    }
                }
            }
        }
        println!("{}: {:?}", line, digits);
        sum += digits.first().unwrap() * 10 + digits.last().unwrap();
    }

    sum
}

fn main() {
    dbg!(part1(SAMPLE1));
    dbg!(part1(DATA));
    dbg!(part2(SAMPLE2));
    dbg!(part2(DATA));
}

#[cfg(test)]
mod tests {

    use super::*;
    use test_case::test_case;

    #[test_case(SAMPLE1, 142)]
    #[test_case(DATA, 55488)]
    fn part1_test(input: &str, expected: u64) {
        let actual = part1(input);
        assert_eq!(actual, expected);
    }

    #[test_case(SAMPLE2, 281)]
    #[test_case(DATA, 55614)]
    fn part2_test(input: &str, expected: u64) {
        let actual = part2(input);
        assert_eq!(actual, expected);
    }
}
